src/Polysemy/Internal/Union.hs

```diff
@@ -131,102 +138,130 @@ type Member e r = MemberNoError e r
 -- | Like 'Member', but will produce an error message if the types are
 -- ambiguous.
 --
--- @since TODO
+-- /Be careful with this./ Due to quirks of 'GHC.TypeLits.TypeError',
+-- the custom error messages emitted by this can potentially override other,
+-- more helpful error messages.
+-- See the discussion in
+-- <https://github.com/polysemy-research/polysemy/issues/227 Issue #227>.
+--
+-- @since 1.2.3.0
 type MemberWithError e r =
-  ( MemberNoError e r
+  ( Find r e
 #ifndef NO_ERROR_MESSAGES
     -- NOTE: The plugin explicitly pattern matches on
-    -- `WhenStuck (IndexOf r _) _`, so if you change this, make sure to change
+    -- `WhenStuck (LocateEffect r _) _`, so if you change this, make sure to change
     -- the corresponding implementation in
     -- Polysemy.Plugin.Fundep.solveBogusError
-  , WhenStuck (IndexOf r (Found r e)) (AmbiguousSend r e)
+  , LocateEffect r e ~ '()
+  , WhenStuck (LocateEffect r e) (AmbiguousSend r e)
 #endif
   )
 
 type MemberNoError e r =
   ( Find r e
-  , e ~ IndexOf r (Found r e)
+#ifndef NO_ERROR_MESSAGES
+  , LocateEffect r e ~ '()
+#endif
   )
 
-
 ------------------------------------------------------------------------------
--- | The kind of type-level natural numbers.
-data Nat = Z | S Nat
-
-
-------------------------------------------------------------------------------
--- | A singleton for 'Nat'.
-data SNat :: Nat -> Type where
-  SZ :: SNat 'Z
-  SS :: SNat n -> SNat ('S n)
-
-instance TestEquality SNat where
-  testEquality SZ     SZ     = Just Refl
-  testEquality (SS _) SZ     = Nothing
-  testEquality SZ     (SS _) = Nothing
-  testEquality (SS n) (SS m) =
-    case testEquality n m of
-      Nothing -> Nothing
-      Just Refl -> Just Refl
-  {-# INLINE testEquality #-}
-
-
-type family IndexOf (ts :: [k]) (n :: Nat) :: k where
-  IndexOf (k ': ks) 'Z = k
-  IndexOf (k ': ks) ('S n) = IndexOf ks n
-
-
-type family Found (ts :: [k]) (t :: k) :: Nat where
+-- | A proof that @e@ is an element of @r@.
+--
+-- Due to technical reasons, @'ElemOf' r e@ is not powerful enough to
+-- prove @'Member' e r@; however, it can still be used send actions of @e@
+-- into @r@ by using 'Polysemy.Internal.subsumeUsing'.
+data ElemOf r e where
+  Here :: ElemOf (e ': r) e
+  In   :: ElemOf r e -> ElemOf (e' ': r) e
+
+instance TestEquality (ElemOf r) where
+  testEquality Here Here
+    = Just Refl
+  testEquality (In e) (In e')
+    = testEquality e e'
+  testEquality _ _ =
+    Nothing
+
+type family LocateEffect (ts :: [k]) (t :: k) :: () where
```

Comment:
What's going on here? Is it just to spit out an error / detect stuckness? The kind here makes me suspicious, since it seems like it should be a no-op. Maybe a comment would help.

Reply:
Yeah, it's just used to detect stuckness -- or rather, signal that an effect is given/detect if an effect isn't given. The old type equality -- `e ~ IndexOf r (Found r e)` -- served two purposes -- to actually make `Member` work, and to detect stuckness. The improved proof covers the former use, so only the latter is relevant anymore, so I simplified the type family to `LocaleEffect`.

Reply:
Added documentation to `LocateEffect` to explain it.